src/components/Collaborative/hooks/useFloatingLinkState.tsx

```diff
@@ -0,0 +1,178 @@
+    );
+
+    const [error, setError] = useState(null);
+    const chain = useChainedCommands();
+    const { isEditing, linkShortcut, setIsEditing, isLoading, setIsLoading } =
+        useLinkShortcut();
+    const { from, to, empty, ranges } = useCurrentSelection();
+    const url = (useAttrs().link()?.href as string) ?? "https://";
+    const [href, setHref] = useState<string>(url);
+
+    const isSelected = useUpdateSelection();
+    const updateReason = useUpdateReason();
+    const floatingLinkPositioner = useMemo(
+        () => createMarkPositioner({ type: "link" }),
+        []
+    );
+
+    useLayoutEffect(() => {
+        if (!isEditing) {
+            return;
+        }
+
+        if (updateReason.doc || updateReason.selection) {
+            setIsEditing(false);
+        }
+    }, [isEditing, setIsEditing, updateReason.doc, updateReason.selection]);
+
+    useEffect(() => setHref(url), [url]);
+
+    const validateFloatingLink = useCallback(() => {
+        const urlPattern = /^(ftp|http|https):\/\/[^ "]+$/;
+        if (!urlPattern.test(href)) {
+            throw new Error(t("sourceForm:errorMessageValidURL"));
+        }
+    }, [href, t]);
+
+    const updateFloatingLink = useCallback(
+        (id) => {
+            const range = linkShortcut ?? undefined;
+
+            if (href === "") {
+                chain.removeLink();
+            } else {
+                chain.updateLink({ href, auto: true, id }, range);
+            }
+            chain.focus(linkShortcut?.to ?? to).run();
+        },
+        [chain, href, linkShortcut, to]
+    );
+
+    const submitHref = useCallback(() => {
+        setIsLoading(true);
+
+        try {
+            const id = uniqueId();
+            updateFloatingLink(id);
+            const { $to } = ranges[0];
+            //@ts-ignore
+            const field = $to?.path[3]?.type?.name;
+            const targetText = $to.doc.textBetween(from, to);
+            const newSource = {
+                href,
+                props: {
+                    field,
+                    targetText,
+                    id,
+                },
+            };
+
+            validateFloatingLink();
+            setEditorSources((sources) => {
+                if (!sources) {
+                    return [newSource];
+                }
+                return [...sources, newSource];
+            });
+        } catch (error) {
+            setError(error.message);
+        } finally {
+            setHref("https://");
+            setIsLoading(false);
+            setIsEditing(false);
+        }
+    }, [
+        setIsLoading,
+        ranges,
+        from,
+        to,
+        href,
+        validateFloatingLink,
+        setEditorSources,
+        updateFloatingLink,
+        setIsEditing,
+    ]);
+
+    const onRemoveLink = useCallback(async () => {
+        setIsLoading(true);
+        setEditorSources((sources) => {
+            const index = editorSources.findIndex(
```

Comment:
todo: create an issue to handle this with ids when dealing with duplicated sources